google/cloud/storage/internal/async/writer_connection_impl.cc

```diff
@@ -0,0 +1,212 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/async/writer_connection_impl.h"
+#include "google/cloud/storage/internal/async/partial_upload.h"
+#include "google/cloud/storage/internal/async/write_payload_impl.h"
+#include "google/cloud/storage/internal/grpc/ctype_cord_workaround.h"
+#include "google/cloud/storage/internal/grpc/object_metadata_parser.h"
+#include "google/cloud/internal/make_status.h"
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+auto HandleFinishAfterError(std::string msg) {
+  return [m = std::move(msg)](future<Status> f) {
+    auto status = f.get();
+    if (!status.ok()) return status;
+    // A `Write()` or `Read()` operation unexpectedly returned `ok == false`.
+    // The `Finish()` call should return the error detailed, it returned
+    // "success". This is some kind of internal error in the client library,
+    // gRPC, or the service.
+    return internal::InternalError(std::move(m), GCP_ERROR_INFO());
+  };
+}
+
+auto HandleFinishAfterError(Status s) {
+  return [s = std::move(s)](future<Status>) mutable { return std::move(s); };
+}
+
+}  // namespace
+
+AsyncWriterConnectionImpl::AsyncWriterConnectionImpl(
+    google::cloud::internal::ImmutableOptions options,
+    std::unique_ptr<StreamingRpc> impl, std::string upload_id,
+    std::shared_ptr<storage::internal::HashFunction> hash_function,
+    std::int64_t persisted_size)
+    : options_(std::move(options)),
+      impl_(std::move(impl)),
+      upload_id_(std::move(upload_id)),
+      hash_function_(std::move(hash_function)),
+      persisted_state_(persisted_size),
+      offset_(persisted_size),
+      finished_(on_finish_.get_future()) {}
+
+AsyncWriterConnectionImpl::AsyncWriterConnectionImpl(
+    google::cloud::internal::ImmutableOptions options,
+    std::unique_ptr<StreamingRpc> impl, std::string upload_id,
+    std::shared_ptr<storage::internal::HashFunction> hash_function,
+    storage::ObjectMetadata metadata)
+    : options_(std::move(options)),
+      impl_(std::move(impl)),
+      upload_id_(std::move(upload_id)),
+      hash_function_(std::move(hash_function)),
+      persisted_state_(std::move(metadata)),
+      finished_(on_finish_.get_future()) {}
+
+AsyncWriterConnectionImpl::~AsyncWriterConnectionImpl() {
+  // Cancel the streaming RPC.
+  impl_->Cancel();
+  // It is safe to call Finish() because:
+  // (1) this is a no-op if it was already called.
+  // (2) calls to `Write()` and `Finalize()` must have completed by the time the
+  //     destructor is called.
+  Finish();
+  // When `impl_->Finish()` is satisfied then `finished_` is satisfied too.
+  // This extends the lifetime of `impl_` until it is safe to delete `impl_`.
+  finished_.then([impl = std::move(impl_)](auto) mutable {  //
+    impl.reset();
```

Comment:
Is the explicit `reset()` important?  If so, consider a comment, and otherwise consider removing it.

Reply:
Yes. Included a comment.